### PR TITLE
Use idtree

### PR DESCRIPTION
**Pull request by agkaminski**
## Description

Replace multiple implementations of interval tree for ID allocations with one universal one.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [x] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: (ia32-generic-qemu).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing linter checks and tests passed.
- [x] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: proc/threads.c [hunk 1]
```diff
@@ -855,6 +749,12 @@ int proc_threadCreate(process_t *process, void (*start)(void *), unsigned int *i
...
 	t->lastTime = t->startTime;
 
+	if (thread_alloc(t) < 0) {
+		vm_kfree(t->kstack);
+		vm_kfree(t);
+		return -ENOMEM;
+	}
+
 	if (process != NULL && (process->tls.tdata_sz != 0 || process->tls.tbss_sz != 0)) {
 		err = process_tlsInit(&t->tls, &process->tls, process->mapp);
```
[the remark this answers is not in the document]
Comment (agkaminski) [reply proc/threads.c:759]: Oops, we should indeed, thx

**Comment by lukileczo** [comment]
One more thing: why the exclamation mark in commit message?

**Review by lukileczo — DISMISSED** [review]
LGTM

File: proc/cond.c [hunk 2]
```diff
@@ -5,115 +5,129 @@
...
 
 
-cond_t *cond_get(unsigned int c)
+void cond_put(cond_t *cond)
 {
-	return resourceof(cond_t, resource, resource_get(proc_current()->process, rtCond, c));
+	thread_t *t = proc_current();
+	int rem = resource_put(t->process, &cond->resource);
+	LIB_ASSERT(rem >= 0, "process: %s, pid: %d, tid: %d, refcnt below zero",
+		t->process->path, process_getPid(t->process), proc_getTid(t));
```
Comment (kemonats) [inline proc/cond.c:40]: What do you think about adding a check if the pointer `cond == NULL` at the begining of this function and removing check before calling this function.
Missing empty line after the declaration of local variables.
```suggestion
	thread_t *t;
	int rem;

	if (cond == NULL) {
		return;
	}
	t = proc_current();
	rem = resource_put(t->process, &cond->resource);
	LIB_ASSERT(rem >= 0, "process: %s, pid: %d, tid: %d, refcnt below zero",
		t->process->path, process_getPid(t->process), proc_getTid(t));
```

Comment (kemonats) [reply proc/cond.c]: But you can remove the check in the `userintr.c` file ( for times ) such as this:
```c
		if (cond != NULL) {
			cond_put(cond);
		}
```

Comment (agkaminski) [reply proc/cond.c]: I'm aware, but it's a matter of - do we allow to `put` a resource you never had? IMHO it's a bad practice to do so and it makes a code using it somewhat less clear

File: proc/threads.c [hunk 3]
```diff
@@ -85,37 +85,14 @@ static int threads_sleepcmp(rbnode_t *n1, rbnode_t *n2)
...
-		return -1;
-
-	else if (t1->id > t2->id)
-		return 1;
-
-	return 0;
+	if (t1->wakeup != t2->wakeup) {
+		return (t1->wakeup > t2->wakeup) ? 1 : -1;
+	}
+	else {
```
Comment (lukileczo) [inline proc/threads.c:91]: Unnecessary `else`

File: proc/cond.c [hunk 4]
```diff
@@ -5,115 +5,134 @@
...
+	if (rem <= 0) {
+		proc_threadBroadcastYield(&cond->queue);
+		vm_kfree(cond);
+	}
 }
 
 
 int proc_condCreate()
```
[the remark this answers is not in the document]
Comment (agkaminski) [reply proc/cond.c]: Fixed

File: proc/threads.c (hunk 3 above)
Comment (agkaminski) [reply proc/threads.c]: Thanks for finding this, fixed

File: proc/threads.c [hunk 5]
```diff
@@ -1128,6 +1023,19 @@ static void _proc_threadDequeue(thread_t *t)
 }
 
 
+static void _proc_threadInsertSleeping(thread_t *t)
```
Comment (lukileczo) [inline proc/threads.c:1026]: Hmm, I'm not convinced that solves the problem: `lib_rbInsert` can only return not `EOK` if `tree->compare` (in this case `threads_sleepcmp`) returns 0 - which it doesn't. So we'd still insert thread with the same `wakeup`

Comment (agkaminski) [reply proc/threads.c]: Right, I've reverted the change

**Review by lukileczo — APPROVED** [review]
LGTM

**Review by jmaksymowicz — APPROVED** [review]
LGTM

**Pull request merged**